test/e2e/upgrade_hypershift_operator_test.go

```diff
@@ -0,0 +1,217 @@
+				nodePoolsMap[nodepools.Items[i].Name] = &nodepools.Items[i]
+			}
+
+			// Get the MachineDeployments
+			machineDeployments := &v1beta1.MachineDeploymentList{}
+			hcpNameSpace = manifests.HostedControlPlaneNamespace(hostedCluster.Namespace, hostedCluster.Name)
+
+			err = mgmtClient.List(ctx, machineDeployments, crclient.InNamespace(hcpNameSpace))
+
+			g.Expect(err).ToNot(gomega.HaveOccurred(),
+				"Listing MachineDeployments in namespace %s shouldn't return errors", hcpNameSpace)
+			g.Expect(machineDeployments.Items).ToNot(gomega.BeEmpty(),
+				"Should find MachineDeployments in namespace %s", hcpNameSpace)
+			g.Expect(len(machineDeployments.Items)).To(gomega.BeEquivalentTo(len(nodepools.Items)),
+				"Number of MachineDeployments and NodePools should match")
+
+			machineDeploymentMap = make(map[string]*v1beta1.MachineDeployment)
+			t.Logf("Found %d MachineDeployments", len(machineDeployments.Items))
+			for i := range machineDeployments.Items {
+				t.Logf("Found MachineDeployment %s", machineDeployments.Items[i].Name)
+				machineDeploymentMap[machineDeployments.Items[i].Name] = &machineDeployments.Items[i]
+			}
+
+		})).To(gomega.BeTrue(), "Calculating HyperShift Operator upgrade invariants should succeed")
+
+		g.Expect(t.Run("Upgrade HyperShift Operator", func(t *testing.T) {
+			t.Logf("Upgrading HyperShift Operator to image %s", hyperShiftOperatorLatestImage)
+			globalOpts.HyperShiftOperatorLatestImage = hyperShiftOperatorLatestImage
+			// Note that we're replacing previous HO installation cleanup function with the new one
+			cleanup, err = e2eutil.InstallHyperShiftOperator(ctx, globalOpts)
+			if err != nil {
+				t.Fatal("Failed to upgrade HyperShift Operator")
+			}
+		})).To(gomega.BeTrue(), "Upgrade HyperShift Operator should succeed")
+
+		g.Expect(t.Run("Verify upgrade invariants", func(t *testing.T) {
+			t.Log("Verifying upgrade invariants")
+			namespace := hostedCluster.Namespace
+			g := gomega.NewWithT(t)
+
+			postUpgradeNodePools := &hyperv1.NodePoolList{}
+			err = mgmtClient.List(ctx, postUpgradeNodePools, crclient.InNamespace(namespace))
+			g.Expect(err).ToNot(gomega.HaveOccurred(), "Listing NodePools in namespace %s shouldn't return errors",
+				namespace)
+
+			g.Expect(postUpgradeNodePools.Items).ToNot(gomega.BeEmpty(), "Should find NodePools in namespace %s",
+				namespace)
+			g.Expect(len(postUpgradeNodePools.Items)).To(gomega.BeEquivalentTo(len(nodePoolsMap)),
+				"Number of pre-ugrade and post-upgrade NodePools should match")
+
+			for _, nodePool := range postUpgradeNodePools.Items {
+				t.Logf("Verifying NodePool %s", nodePool.Name)
+				preUpgradeNodePool, ok := nodePoolsMap[nodePool.Name]
+				g.Expect(ok).To(gomega.BeTrue(), "Should have a pre-upgrade nodepool with the name %s",
+					nodePool.Name)
+
+				t.Logf("Generation: %d", nodePool.Generation)
+				t.Logf("CurrentConfig: %s", nodePool.Annotations[nodePoolAnnotationCurrentConfig])
```

Comment:
can we also check the relevant conditions?
```
NodePoolUpdatingVersionConditionType = "UpdatingVersion"
	// NodePoolUpdatingConfigConditionType signals if a config update is currently happening in NodePool.
	NodePoolUpdatingConfigConditionType = "UpdatingConfig"
	// NodePoolUpdatingPlatformMachineTemplateConditionType signals if a platform machine template update is currently happening in NodePool.
	NodePoolUpdatingPlatformMachineTemplateConditionType = "UpdatingPlatformMachineTemplate"
```

Reply:
Done